### PR TITLE
GH-101696: invalidate type version tag in `_PyStaticType_Dealloc `

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-101696

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2023-02-08-17-13-31.gh-issue-101696.seJhTt.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Invalidate type version tag in ``_PyStaticType_Dealloc`` for static types avoiding bug where a false cache hit could crash the interpreter. Patch by Kumar Aditya
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Core and Builtins/2023-02-08-17-13-31.gh-issue-101696.seJhTt.rst]:
```suggestion
Invalidate type version tag in ``_PyStaticType_Dealloc`` for static types, avoiding bug where a false cache hit could crash the interpreter. Patch by Kumar Aditya.
```

**Review by erlend-aasland — APPROVED** [review]

**Comment by markshannon** [comment]
This makes no sense to me.

Static types should be immutable; their version number should never change.

See https://github.com/python/cpython/issues/101603

**Comment by kumaraditya303** [comment]
> Static types should be immutable; their version number should never change.

It is still the case but when the interpreter is initialized multiple times then the type still has the same old version tag and can cause assertion failures. Note that the type cache is per interp now and is cleared on every cycle. See https://github.com/python/cpython/actions/runs/4112993949/jobs/7098592004 as an example.

**Comment by markshannon** [comment]
I don't think that this PR fixes the root problem, although it almost certainly no worse than what we had before.

What assertions fails? "Can cause assertion failures" doesn't tell me anything.

When you say "the interpreter" how does this handle sub-interpreters?
The problems are that heap and static types are sharing the same versions, and that versions are shared across interpreters.

@ericsnowcurrently might have thoughts on this.

**Comment by erlend-aasland** [comment]
> What assertions fails? "Can cause assertion failures" doesn't tell me anything.

See the GH Action run that Kumar linked to for an example. I'll quote it for you:

```
Objects/typeobject.c:4174: _PyType_Lookup: Assertion `_PyType_HasFeature(type, Py_TPFLAGS_VALID_VERSION_TAG)' failed.
```

This is from `test_finalize_structseq` in `test_embed`.

**Pull request merged**